### PR TITLE
ISAICP-5454: Bcc moderators on solution changes request.

### PR DESCRIPTION
**Pull request by idimopoulos**

### CODE REVIEW

File: web/modules/custom/joinup_notification/src/JoinupMessageDeliveryInterface.php [hunk 1]
```diff
@@ -138,6 +138,19 @@ public function setRecipients(array $accounts): self;
...
   public function setRecipientsAsEmails(array $mails): self;
 
+  /**
+   * Sets the list of accounts that will be added to bcc.
+   *
+   * @param array $bcc_emails
+   *   A list of user emails.
+   *
+   * @return $this
+   *   Return the object for chaining.
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/JoinupMessageDeliveryInterface.php]: `@return $this` is a special kind of fish that doesn't need a comment. From https://www.drupal.org/docs/develop/standards/api-documentation-and-comment-standards#types

> You may omit the description if using `@return $this` or `@return static`.

However, it's not wrong, you can keep it.

File: web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php [hunk 2]
```diff
@@ -204,6 +205,49 @@ protected function getUsersMessages(array $user_data, EntityInterface $entity =
...
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The group entity.
+   * @param array $bcc_data
+   *   An associative array of roles indexed by the 'roles' key or the
+   *   'og_roles' key depending on whether it is a site wide role or an og role.
+   * @param array $uids_to_skip
+   *   (optional) An array of ids to skip.
+   *
+   * @return array
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php]: s/array/string[]

File: web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php [hunk 3]
```diff
@@ -313,12 +357,14 @@ protected function generateArguments(EntityInterface $entity) {
...
-   *   Optionally pass additional arguments.
+   *   (optional) Additional arguments to be passed to the message.
+   * @param array $bcc_emails
+   *   (optional) A list of emails to be passed as Bcc.
    *
    * @return bool
    *   Whether or not the messages were sent successfully.
    */
-  protected function sendUserDataMessages(array $user_data, array $arguments = []) : bool {
+  protected function sendUserDataMessages(array $user_data, array $arguments = [], array $bcc_emails = []) : bool {
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php]: `) : bool` -> `): bool`

File: web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php [hunk 4]
```diff
@@ -313,12 +357,14 @@ protected function generateArguments(EntityInterface $entity) {
    * @param array $user_data
    *   An array of user ids and their corresponding messages.
    * @param array $arguments
-   *   Optionally pass additional arguments.
+   *   (optional) Additional arguments to be passed to the message.
+   * @param array $bcc_emails
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php]: s/array/string[]

File: web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php [hunk 5]
```diff
@@ -204,6 +205,49 @@ protected function getUsersMessages(array $user_data, EntityInterface $entity =
...
 
+  /**
+   * Retrieves a list of emails from a given list of roles.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The group entity.
+   * @param array $bcc_data
+   *   An associative array of roles indexed by the 'roles' key or the
+   *   'og_roles' key depending on whether it is a site wide role or an og role.
+   * @param array $uids_to_skip
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php]: s/array/int[]

File: web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php [hunk 6]
```diff
@@ -204,6 +205,49 @@ protected function getUsersMessages(array $user_data, EntityInterface $entity =
...
+   *   An array of emails.
+   */
+  protected function getBccEmails(EntityInterface $entity, array $bcc_data, array $uids_to_skip = []): array {
+    $return = [];
+    foreach ($bcc_data['roles'] as $role_id) {
+      $uids = $this->getRecipientIdsByRole($role_id);
+      if ($uids = array_diff(array_values($uids), $uids_to_skip)) {
+        $emails = array_map(function (UserInterface $user): string {
+          return $user->getEmail();
+        }, $this->entityTypeManager->getStorage('user')->loadMultiple($uids));
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php]: `$this->entityTypeManager->getStorage('user')` is computed on each iteration. Why not setting a var?

File: web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php [hunk 7]
```diff
@@ -204,6 +205,49 @@ protected function getUsersMessages(array $user_data, EntityInterface $entity =
...
+
+    if (isset($bcc_data['og_roles'])) {
+      $og_role_storage = $this->entityTypeManager->getStorage('og_role');
+      foreach ($bcc_data['og_roles'] as $role_id => $messages) {
+        $role = $og_role_storage->load($role_id);
+        $uids = $this->getRecipientIdsByOgRole($entity, $role);
+        if ($uids = array_diff(array_values($uids), $uids_to_skip)) {
+          $emails = array_map(function (UserInterface $user): string {
+            return $user->getEmail();
+          }, $this->entityTypeManager->getStorage('user')->loadMultiple($uids));
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php]: Same. Use the var.

File: web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php [hunk 8]
```diff
@@ -204,6 +205,49 @@ protected function getUsersMessages(array $user_data, EntityInterface $entity =
...
+          return $user->getEmail();
+        }, $this->entityTypeManager->getStorage('user')->loadMultiple($uids));
+        $return += $emails;
+      }
+    }
+
+    if (isset($bcc_data['og_roles'])) {
+      $og_role_storage = $this->entityTypeManager->getStorage('og_role');
+      foreach ($bcc_data['og_roles'] as $role_id => $messages) {
+        $role = $og_role_storage->load($role_id);
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_notification/src/EventSubscriber/NotificationSubscriberBase.php]: Why not loading all once with `loadMultiple()`?

File: web/modules/custom/contact_form/src/EventSubscriber/NotificationSubscriber.php [hunk 9]
```diff
@@ -127,13 +127,14 @@ protected function sendReportMessage(NotificationEvent $event) {
    *
    * Skip generating the arguments during the sending process.
    */
-  protected function sendUserDataMessages(array $user_data, array $arguments = []) : bool {
+  protected function sendUserDataMessages(array $user_data, array $arguments = [], array $bcc_emails = []) : bool {
```
Comment (claudiu-cristea) [inline web/modules/custom/contact_form/src/EventSubscriber/NotificationSubscriber.php]: `) : bool` -> `): bool`

**Comment by idimopoulos** [comment]
I agree with the changes.

**Pull request merged**